cmd/command_argument_validator.go

```diff
@@ -0,0 +1,25 @@
+package cmd
+
+import "fmt"
+
+type CommandArgumentValidator struct {
+	required int
+	optional int
+}
+
+func (c *CommandArgumentValidator) validate(args []string) (err error) {
+	argCount := len(args)
+
+	expectedCount := fmt.Sprintf("exactly %d", c.required)
+	if (c.optional > 0) {
+		expectedCount = fmt.Sprintf("between %d and %d", c.required, c.required+c.optional)
+	}
+
+	if argCount > c.required+c.optional {
+		err = fmt.Errorf("Error: too many arguments (expected %s, got %d)\n", expectedCount, len(args))
+	} else if argCount < c.required {
+		err = fmt.Errorf("Error: missing arguments (expected %s, got %d)\n", expectedCount, len(args))
+	}
+
+	return
```

Comment:
🤔 does it complain if you explicitly do `return err`? Makes more sense to me if not (even though this works)

Reply:
- added explicitly return in case https://github.com/golang/go/issues/21291 happens
- return inside the if block to short-circuit it